pkg/receive/handler.go

```diff
@@ -0,0 +1,222 @@
+package receive
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	stdlog "log"
+	"net"
+	"net/http"
+	"runtime/debug"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cmux"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/improbable-eng/thanos/pkg/store/prompb"
+	conntrack "github.com/mwitkow/go-conntrack"
+	"github.com/opentracing-contrib/go-stdlib/nethttp"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/route"
+	"github.com/prometheus/prometheus/storage/tsdb"
+	"golang.org/x/net/netutil"
+)
+
+var (
+	requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "prometheus_http_request_duration_seconds",
+			Help:    "Histogram of latencies for HTTP requests.",
+			Buckets: []float64{.1, .2, .4, 1, 3, 8, 20, 60, 120},
+		},
+		[]string{"handler"},
+	)
+	responseSize = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "prometheus_http_response_size_bytes",
+			Help:    "Histogram of response size for HTTP requests.",
+			Buckets: prometheus.ExponentialBuckets(100, 10, 8),
+		},
+		[]string{"handler"},
+	)
+)
+
+// Options for the web Handler.
+type Options struct {
+	Receiver       *ReceiveWriter
+	Context        context.Context
+	ListenAddress  string
+	MaxConnections int
+	ReadTimeout    time.Duration
+	Registry       prometheus.Registerer
+	ReadyStorage   *tsdb.ReadyStorage
+}
+
+// Handler serves various HTTP endpoints of the Prometheus server
```

Comment:
Missing trailing period. Also `various` is kind of wide word..  I think we expose only `receive` right? (: